test/lit/passes/optimize-instructions_branch-hints-fold.wast

```diff
@@ -0,0 +1,220 @@
+ ;; NO_FO-NEXT:    (i32.const 42)
+ ;; NO_FO-NEXT:   )
+ ;; NO_FO-NEXT:  )
+ ;; NO_FO-NEXT: )
+ (func $conditionals (param $x i32) (result i32)
+  ;; When we flip the if, the hint should flip too.
+  (@metadata.code.branch_hint "\00")
+  (if (result i32)
+   (i32.eqz
+    (local.get $x)
+   )
+   (then
+    (i32.const 42)
+   )
+   (else
+    (i32.const 1337)
+   )
+  )
+ )
+
+ ;; CHECK:      (func $still-fold (type $0) (param $x i32) (param $y i32)
+ ;; CHECK-NEXT:  (@metadata.code.branch_hint "\00")
+ ;; CHECK-NEXT:  (if
+ ;; CHECK-NEXT:   (local.get $y)
+ ;; CHECK-NEXT:   (then
+ ;; CHECK-NEXT:    (unreachable)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ ;; NO_FO:      (func $still-fold (type $0) (param $x i32) (param $y i32)
+ ;; NO_FO-NEXT:  (if
+ ;; NO_FO-NEXT:   (local.get $x)
+ ;; NO_FO-NEXT:   (then
+ ;; NO_FO-NEXT:    (@metadata.code.branch_hint "\00")
+ ;; NO_FO-NEXT:    (if
+ ;; NO_FO-NEXT:     (local.get $y)
+ ;; NO_FO-NEXT:     (then
+ ;; NO_FO-NEXT:      (unreachable)
+ ;; NO_FO-NEXT:     )
+ ;; NO_FO-NEXT:    )
+ ;; NO_FO-NEXT:   )
+ ;; NO_FO-NEXT:   (else
+ ;; NO_FO-NEXT:    (@metadata.code.branch_hint "\01")
+ ;; NO_FO-NEXT:    (if
+ ;; NO_FO-NEXT:     (local.get $y)
+ ;; NO_FO-NEXT:     (then
+ ;; NO_FO-NEXT:      (unreachable)
+ ;; NO_FO-NEXT:     )
+ ;; NO_FO-NEXT:    )
+ ;; NO_FO-NEXT:   )
+ ;; NO_FO-NEXT:  )
+ ;; NO_FO-NEXT: )
+ (func $still-fold (param $x i32) (param $y i32)
+  ;; We fold if arms even if metadata differs (like LLVM). We do not fold if the
+  ;; flag was passed, however.
+  (if
+   (local.get $x)
+   (then
+    (@metadata.code.branch_hint "\00")
+    (if
+     (local.get $y)
+     (then
+      (unreachable)
+     )
+    )
+   )
+   (else
+    (@metadata.code.branch_hint "\01")
+    (if
+     (local.get $y)
+     (then
+      (unreachable)
+     )
+    )
+   )
+  )
+ )
+
+ ;; CHECK:      (func $yes-fold (type $0) (param $x i32) (param $y i32)
+ ;; CHECK-NEXT:  (@metadata.code.branch_hint "\01")
+ ;; CHECK-NEXT:  (if
+ ;; CHECK-NEXT:   (local.get $y)
+ ;; CHECK-NEXT:   (then
+ ;; CHECK-NEXT:    (unreachable)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ ;; NO_FO:      (func $yes-fold (type $0) (param $x i32) (param $y i32)
+ ;; NO_FO-NEXT:  (if
+ ;; NO_FO-NEXT:   (local.get $x)
+ ;; NO_FO-NEXT:   (then
+ ;; NO_FO-NEXT:    (@metadata.code.branch_hint "\01")
+ ;; NO_FO-NEXT:    (if
+ ;; NO_FO-NEXT:     (local.get $y)
+ ;; NO_FO-NEXT:     (then
+ ;; NO_FO-NEXT:      (unreachable)
+ ;; NO_FO-NEXT:     )
+ ;; NO_FO-NEXT:    )
+ ;; NO_FO-NEXT:   )
+ ;; NO_FO-NEXT:   (else
+ ;; NO_FO-NEXT:    (@metadata.code.branch_hint "\01")
+ ;; NO_FO-NEXT:    (if
+ ;; NO_FO-NEXT:     (local.get $y)
+ ;; NO_FO-NEXT:     (then
+ ;; NO_FO-NEXT:      (unreachable)
+ ;; NO_FO-NEXT:     )
+ ;; NO_FO-NEXT:    )
+ ;; NO_FO-NEXT:   )
+ ;; NO_FO-NEXT:  )
+ ;; NO_FO-NEXT: )
+ (func $yes-fold (param $x i32) (param $y i32)
+  ;; Now the hints match, so we definitely fold (without the flag).
```

Comment:
Would it be worth having the flag only disable folding when the metadata is different?

Reply:
I guess that would show the fuzzer finds no issues when folding identical metadata, but I'm not sure it's worth the complexity in the code to be that precise. We know that is safe to do, so it would only be checking that we wrote that condition on the flag properly.